src/io.c

```diff
@@ -417,3 +417,41 @@ niova_io_iov_restore(struct iovec *iovs, size_t niovs, size_t save_idx,
 
     return 0;
 }
+
+/**
+ * iovec iterator - calls callback once per iovec.
+ *
+ * @iovs:      Array of iovecs containing the data
+ * @niovs:     Number of iovecs in the array
+ * @num_bytes: Number of bytes to iterate over
+ * @cb:        Callback function invoked for each iovec segment
+ * @user_ctx:  Context pointer passed through to callback
+ *
+ */
+ssize_t
+niova_io_iterate_iovs(const struct iovec *iovs, size_t niovs,
+                      size_t num_bytes, niova_iov_cb_t cb, void *user_ctx)
+{
+    if (!iovs || !niovs || !cb)
+        return -EINVAL;
+
+    if (num_bytes == 0)
+        return 0;
+
+    size_t bytes_remaining = num_bytes;
+
+    for (size_t iov_idx = 0; bytes_remaining > 0 && iov_idx < niovs; iov_idx++)
+    {
+        size_t segment_len = MIN(iovs[iov_idx].iov_len, bytes_remaining);
+        int cb_rc = cb(iovs[iov_idx].iov_base, segment_len, user_ctx);
+        if (cb_rc != 0)
+            return cb_rc;
+
+        bytes_remaining -= segment_len;
+    }
+
+    if (bytes_remaining > 0)
+        return -ERANGE;
+
+    return (ssize_t)(num_bytes - bytes_remaining);
```

Comment:
It's good to return the value of "total bytes processed" but in this current code, `bytes_remaining` will always be `0`.   In addition, we may want to consider `num_bytes` of 0 being allowed. What I suggest is that you keep a local count of processed bytes and return that rather calculating it.